### PR TITLE
Modernise grammar and fix bugs

### LINKED ISSUE #280: Missing scopes on periods
Periods aren't consistently scoped:

```javascript
var result = object.method();       // Correct: meta.delimiter.method.period.js

var instance = new object.Class();  // Incorrect: entity.name.type.instance.js


object.property = value;            // Correct: meta.delimiter.method.period.js

object.method = function(){};       // Incorrect: meta.function.js
```

k

### PR DESCRIPTION
**Pull request by Alhadis**
There's quite a bit happening with this PR, so hopefully my explanations are digestible.

Added: Support for ES6 Unicode escapes
-------------------
Figure 1

Fixed: Bugs with constructors
-----------------------------
Constructors containing dollar-signs or multiple constructors weren't matched the same way as those who didn't (also fixes #280):

Figure 2

Fixed: Issues with dollar-signs
------------
Fixes [#409: Wrong highlighting for constants with dollar sign](https://github.com/atom/language-javascript/issues/409):
```js
ANGULAR
ANG$$$$$ULAR
```

Also fixed an issue I picked up with constants not matching if they begin with an underscore:

```js
_CONSTANT
 CONSTANT
```

Also fixes [#344: Wrong scope on `$this` and `$default`](https://github.com/atom/language-javascript/issues/344):
```js
 this
$this
abc$this
```

Fixed: "Support constants" only recognised without space
--------------------------------------------------------
Minor fix:

```js
window.innerWidth
window.  innerWidth
```

Fixed: Missing highlighting for ES6+ classes
--------------------------------------------
Added highlighting for modern [JavaScript classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference) added in recent versions of the language:

```js
ArrayBuffer
Atomics
DataView
Error
EvalError
Float32Array
Float64Array
Generator
GeneratorFunction
Int16Array
Int32Array
Int8Array
InternalError
Intl
Promise
Proxy
RangeError
ReferenceError
Reflect
SharedArrayBuffer
SIMD
SyntaxError
TypeError
Uint16Array
Uint32Array
Uint8Array
Uint8ClampedArray
URIError
```

Added: Better DOM support
-------------------------
I [wrote a script](https://gist.github.com/Alhadis/16effd7a22522c793a41bbc65424eed8) to generate a regex block from the [web interfaces documented on MDN](https://developer.mozilla.org/en-US/docs/Web/API). Note I purposefully omitted any APIs that're deprecated, non-standard, or still considered "experimental".

At the same time, I've also removed support for these keywords, which refer to Netscape-era technologies that've been deprecated since forever:

```js
// Removed properties:
arity
java
sun
netscape

// Removed classes:
Anchor
Applet
Area
Button
Checkbox
FileUpload
Form
Frame
Hidden
Image
JavaArray
JavaClass
JavaObject
JavaPackage
Layer
Link
MimeType
Option
Packages
Password
Radio
Reset
Select
Style
Submit
Textarea
```

The latter list has been superseded by standardised HTML* classes. For example:

* `Anchor` -> [`HTMLAnchor`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLAnchorElement)
* `Area` -> [`HTMLArea`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLAreaElement)
* `Select` -> [`HTMLSelect`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement)

### CODE REVIEW

**Comment by MaximSokolov** [comment]
Also, there is a conflict with #408

**Comment by Alhadis** [comment]
Not anymore.

**Comment by Alhadis** [comment]
Aaaaand suddenly we have failing specs.



... hang on, fixing.

**Comment by Alhadis** [comment]
Okay, done. @MaximSokolov, if you need to make any further changes, remember to rebase against my branch.

**Comment by 50Wliu** [comment]
This is including some changes from #408 - could you please rebase against master @Alhadis?

**Comment by Alhadis** [comment]
[Already did](https://github.com/atom/language-javascript/pull/415/commits/b9ba47bde6a7491220f65824ac0b8e1e9f71afd0).

File: grammars/javascript.cson [hunk 1]
```diff
@@ -705,13 +705,19 @@
     'name': 'meta.class.js'
   }
   {
+    'match': '(new)\\s+([\\w$]+[\\w.$]*)'
+    'name': 'meta.class.instance.constructor'
```
Comment (50Wliu) [inline grammars/javascript.cson]: Can you add a `.js` to the end of this?

File: grammars/javascript.cson [hunk 2]
```diff
@@ -705,13 +705,19 @@
...
     'captures':
       '1':
         'name': 'keyword.operator.new.js'
       '2':
         'name': 'entity.name.type.instance.js'
-    'match': '(new)\\s+(\\$?\\w+(?:\\.\\w*)?)'
-    'name': 'meta.class.instance.constructor'
+        'patterns':[
```
Comment (50Wliu) [inline grammars/javascript.cson]: Missing space after colon

**Comment by 50Wliu** [comment]
Please keep the specs formatting in line with the existing ones.

**Comment by Alhadis** [comment]
What do you mean? Aside from the two things you just pointed out.

**Comment by Alhadis** [comment]
I'm guessing you're referring to the columnar-layout of the "Unicode escape sequences" block, right? Fixed.

File: spec/javascript-spec.coffee [hunk 3]
```diff
@@ -88,19 +141,37 @@ describe "Javascript grammar", ->
...
       {tokens} = grammar.tokenizeLine('new $something')
-      expect(tokens[0]).toEqual value: 'new', scopes: ['source.js', 'meta.class.instance.constructor', 'keyword.operator.new.js']
-      expect(tokens[1]).toEqual value: ' ', scopes: ['source.js', 'meta.class.instance.constructor']
-      expect(tokens[2]).toEqual value: '$something', scopes: ['source.js', 'meta.class.instance.constructor', 'entity.name.type.instance.js']
+      expect(tokens[0]).toEqual value: 'new', scopes: ['source.js', 'meta.class.instance.constructor.js', 'keyword.operator.new.js']
+      expect(tokens[1]).toEqual value: ' ', scopes: ['source.js', 'meta.class.instance.constructor.js']
+      expect(tokens[2]).toEqual value: '$something', scopes: ['source.js', 'meta.class.instance.constructor.js', 'entity.name.type.instance.js']
+
+      {tokens} = grammar.tokenizeLine('var instance = new obj.ct.Cla$s();')
+      expect(tokens).toEqual [
```
Comment (50Wliu) [inline spec/javascript-spec.coffee]: I'm also talking about these specs

File: spec/javascript-spec.coffee [hunk 4]
```diff
@@ -1138,6 +1220,12 @@ describe "Javascript grammar", ->
       {tokens} = grammar.tokenizeLine('this.obj.prototype = new El()')
       expect(tokens[0]).toEqual value: 'this', scopes: ['source.js', 'variable.language.js']
 
+      {tokens} = grammar.tokenizeLine('$this')
+      expect(tokens[0].value).toEqual '$this'
```
Comment (50Wliu) [inline spec/javascript-spec.coffee]: and these

Comment (Alhadis) [inline spec/javascript-spec.coffee]: @50Wliu Can you explain to me *exactly* what it is you want me to do? I'm really not following here.

Comment (50Wliu) [inline spec/javascript-spec.coffee]: Keep them in `expect(tokens[x]).toEqual value: 'value', scopes: ['scopes']` form.

Comment (50Wliu) [inline spec/javascript-spec.coffee]: Though I guess you don't really have to change this one, but definitely the one on L159.

Comment (Alhadis) [inline spec/javascript-spec.coffee]: Changed.

**Pull request merged**